src/stdlib_math_diff.fypp

```diff
@@ -0,0 +1,71 @@
+#! Originally written in 2016 by Keurfon Luu (keurfonluu@outlook.com)
+#! https://github.com/keurfonluu/Forlab/blob/master/src/lib/forlab.f90#L2673
+
+#:include "common.fypp"
+#:set RI_KINDS_TYPES = REAL_KINDS_TYPES + INT_KINDS_TYPES
+submodule (stdlib_math) stdlib_math_diff
+
+    implicit none
+
+contains
+
+    #! `diff` computes differences of arrays of the ${t1}$ type.
+    
+    #:for k1, t1 in RI_KINDS_TYPES
+    pure module function diff_1_${k1}$(X, n) result(Y)
+        ${t1}$, intent(in) :: X(:)
```

Comment:
Not sure if this interface has been discussed anywhere so far? A few potential issues that come to mind are:
* This routine allocates its output on every call. I can imagine in some situations that could be inefficient. Users might wish to make a more efficient `diff` call by providing the output data. For that case a subroutine interface may be preferable.
* When computing differences in practice, there are different approaches that may be desired to treat the end-points. The current routines reduce the size of the array. But one might also wish to preserve the array size and do some kind of extrapolation at the missing entry. There can be a number of cases here (e.g. forward differences, backward differences, centred differences, with different implications for end-points).

Currently the routine won't support such things. While that might prove to be fine, I'd like to see some discussion to ensure we've chosen the right interface interface.